[SNAP-338] avoid pre-compilation of meta-data procedures, optimize hive meta-store

sumwale opened the pull request:
## Changes proposed in this pull request

- commented out the code to pre-compile the system procedures for meta-data calls
  which was added due to deadlock scenarios with DataDictionary locks
- now compile system procedures like normal procedures avoiding any special DataDictionary
  entries that were the source of problems culminating in the addition of pre-compilation

Move catalog initialization and consistency check to background

- now avoid taking the DLock in catalog initialization in store boot rather is to be done by
  SnappyHiveCatalog init in its own thread
- submit the task for consistency check rather than doing in foreground; metastore reads
   will wait on initialization to complete (if still ongoing)
- removed dependency of gemfire-shared on spark-unsafe (will be used for stand-alone
  launcher in subsequent PRs for SNAP-338)
- added SKIP_HIVE_TABLE_CALLS in HiveTablesVTI to skip hive table lookup from inside the
  hive-metastore thread itself

## Patch testing

store precheckin

## ReleaseNotes changes

NA

## Other PRs 

https://github.com/SnappyDataInc/snappydata/pull/911

Commit:
[SNAP-338] speed up locator startup

- changed default p2p.discoveryTimeout to 3s (from 5s) and sleep between
  join retries to 1s (from 2s); these also match geode defaults:
    https://github.com/apache/geode/blob/develop/geode-core/src/main/java/org/apache/geode/distributed/internal/membership/gms/membership/GMSJoinLeave.java#L97
- added a "p2p.minJoinRetries" property which can be used to adjust the retries done
  in discovering other members but always results in an unncessary 3-4s sleep wait by
  first locator; this is retained as 2 by default as before but will be reduced to 1
  in default SnappyData scripts for local cluster
- removed send of old gossip version by locator which was used by ancient GemFire releases that
  are no longer supported; it caused an exception on receiver (self-join) which would cause further delay
- also reduced the join waits for ConnectionSignaller and SnappyThriftServer during shutdown
  (else are force shut in any case)

Commit:
[SNAP-338] use closure instead of class reflection for serializer

- avoid unnecessarily loading all the serializable classes which is about the most expensive portion
  of node startup (all of locator, server, lead); using JDK8's lamba with Supplier interface instead
  to pass through closures that avoid loading class at registration and is also faster than reflection;
  many classes are also never used in SnappyData (e.g. admin API); updated the GFXD side registration
- also simplifies the special handling for GFXD "fixed instance" which are singletons since the
  closure can encapsulate returning the singleton; removed that and subsumed in the main map
- added a base implementation for NativeCalls.isProcessActive for "pure" mode using /proc

Also improved the spin wait in SystemFailure watchDog and proctor threads to reduce blocking
on them in shutdown:

- moved the thread wait of watchDog outside of the sync block that notifies else it
  results stopped just holding the lock while waiting on thread preventing the
  notified watchDog thread from ending its wait
- similar fix for the proctor thread

Commit:
[SNAP-338] don't pre-compile meta-data procedures

- commented out the code to pre-compile the system procedures for meta-data calls
  which was added due to deadlock scenarios with DataDictionary locks
- now compile system procedures like normal procedures avoiding any special DataDictionary
  entries that were the source of problems culminating in the addition of pre-compilation

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-338_3

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-338

Commit:
Merge remote-tracking branch 'origin/SNAP-338' into SNAP-338_2

Commit:
Merge remote-tracking branch 'origin/SNAP-338_2' into SNAP-338_3

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-338

Commit:
Merge remote-tracking branch 'origin/SNAP-338' into SNAP-338_2

Commit:
Merge remote-tracking branch 'origin/SNAP-338_2' into SNAP-338_3

Commit:
move catalog initialization and consistency check to background

- now avoid taking the DLock in catalog initialization in store boot rather
  is to be done by SnappyHiveCatalog init in its own thread
- submit the task for consistency check rather than doing in foreground
- removed dependency of gemfire-shared on spark-unsafe (will be used for stand-alone launcher)

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-338

Commit:
Merge remote-tracking branch 'origin/SNAP-338' into SNAP-338_2

Commit:
Merge remote-tracking branch 'origin/SNAP-338_2' into SNAP-338_3

Commit:
added SKIP_HIVE_TABLE_CALLS in HiveTablesVTI

this is to skip hive table lookup from inside the hive-metastore thread itself

Commit:
a pending change to ConnectionTable

Commit:
fixed an NPE in catalog due to delayed initialization

Commit:
fix a bug in ProcedureTestDUnit.verifyDataNodeDapStats [SNAP-1667]

fixes occasional failures in ProcedureTestDUnit.testDAPRelatedStatsOnTable

Commit:
fixing occasional failure in SnapshotTransactionDUnit.testRollbackDueToConflictRestart

Commit:
Couple more waiting points in recovery with data removed

- publish initial tableStats from hive metastore client initialization thread
  and not in post DDL replay to avoid blocking boot thread on hive metastore initialization
- correct the monitor used for signaling from index BatchJobControl/BatchJob; it was timing
  out on wait always for all indexes instead of proper signaling

Commit:
increased max time to wait for hive metastore initialization to 60s

Commit:
fix catalog consistency check thread waiting on itself

- skip waiting for catalog consistency thread for a call within itself causing hang
- fixed more occasional failures in ExpirationDUnit

Commit:
publish column stats after catalog consistency check

kneeraj commented on gemfire-shared/src/main/java/com/gemstone/gemfire/internal/shared/SystemProperties.java:
```diff
@@ -228,7 +228,8 @@ public static SystemProperties getClientInstance() {
     try {
-      Class<?> factoryProvider = Class.forName(GFXD_FACTORY_PROVIDER);
+      Class<?> factoryProvider = Class.forName(GFXD_FACTORY_PROVIDER,
+          true, ClassLoader.getSystemClassLoader());
```
Why was this required. Any other class loader getting used instead of system class loader? Just to understand.

sumwale replied:
Yes this change has been reverted in next PR https://github.com/SnappyDataInc/snappy-store/pull/347/files#diff-834f00639cc7c1a04f4d4f8455251e42

Had added this because was testing a custom system loader that had a faster indexed class lookup from jars. However, dropped it because advantage was not significant (~1s for each node).

kneeraj reviewed (commented):
LGTM

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-338_3

Commit:
avoid premature wrapping of SparkException

fixes failures in snappy dunit security tests that expect a different SQLState in the failure

Commit:
removed explicit system classloader usage in couple of places

not required and incorrect when running with Spark

Commit:
fixing failure in MemoryAnalyticsTest

Commit:
fixing unit test failures after the master changes

master changed the way disk paths are evaluated (using realpath) so updating tests for the same
